proposals/0512-nofieldselectors-per-datatype.md

```diff
@@ -0,0 +1,224 @@
+Being able to control field selector generation on a per-datatype level and per-field level lets you use this trick while keeping other "normal" records the same.
+
+### 1.1.3 Consistency with `OverlappingInstances`
+
+The `OverlappingInstances` pragma is used to allow *all* instances in a module to be overlapping or overlappable.
+But this is too coarse grained - you usually want to specify a *single* instance as `{-# OVERLAPPABLE #-}` directly in the instance body.
+This proposal is motivated in a similar manner to specializing `OverlappingInstances` to specific places where you want it.
+
+## 1.2 Proposed Change Specification
+
+* Export the following data type from a new module `ghc-experimental:GHC.Modifiers.Experimental`:
+
+  ```haskell
+  data FieldSelectorsModifier = NoFieldSelectors | FieldSelectors
+  ```
+
+* Allow the use of ``FieldSelectorsModifier`` as a modifier at the level of
+  datatype declarations, data constructor declarations, or record field
+  declarations.
+
+* For each record field, determine whether to generate a top-level field selector as follows:
+
+  1. consult the value of the ``FieldSelectorModifier`` at the field level if present; otherwise
+  2. consult the value of the ``FieldSelectorModifier`` at the data constructor level if present; otherwise
+  3. consult the value of the ``FieldSelectorModifier`` at the data declaration level if present; otherwise
+  4. check whether the ``FieldSelectors`` language extension is enabled in the current module.
+
+  In case of conflicting modifiers, the first takes priority (ordered by the list above).
+
+## 1.3 Examples
+
+Suppose we had a module that did not enable the `NoFieldSelectors` extension.
+
+```haskell
+{-# language FieldSelectors #-}
+
+module Foo where
+
+-- This datatype does not have field selectors generated
+%NoFieldSelectors
+data Foo = Foo { name :: String
+               , age :: Int }
+```
+
+Meanwhile, with a `%FieldSelectors` annotation, we would be able to override the module behavior on a per datatype basis.
+
+```haskell
+{-# language NoFieldSelectors #-}
+
+module Point where
+
+-- The following declaration has top-level field selector functions
+%FieldSelectors
+data Point = Point { x :: Int
+                   , y :: Int }
+
+-- This does not have field selectors
+data MkPoint = MkPoint { x :: Int
+                       , y :: Int }
+```
+
+The pragma may apply to a single constructor in a sum type.
+
+```haskell
+module OnlyOneSelector where
+
+data CoolSumType 
+    = HasFieldSelectors { x :: Int, y :: Int }
+    | %NoFieldSelectors Doesn'tHaveSelectors { a :: Char, b :: Char }
+```
+
+Field level annotations are also possible.
+
+```haskell
+{-# language FieldSelectors #-}
+
+module Foo where
+
+-- This datatype does not have the `secretInformation` field selectors generated
+-- but the other two are fine
+data Foo = Foo { name :: String
+               , age :: Int
+               , %NoFieldSelectors secretInformation :: [String] }
```

Comment:
```suggestion
               , secretInformation %NoFieldSelectors :: [String] }
```

This was an oversight on my part. Fields are the only context where modifiers are written after, not before the modified entity.